File: python/solver.py
from api import API
from graph import Graph
from timer_class import Timer


def solver(address: str | None, maze: str | None) -> None:
    timer_main = Timer()
    api = API(address, maze)
    graph = Graph()

    print("1 - Fazendo chamada inicial (URL: %s | Labirinto: %s)" % (api.url, api.maze))
    indice_inicial = api.iniciar(graph.vertices)

    print("2 - Explorando o labirinto (API) com o DFS")
    graph.dfs(api, indice_inicial, (-1), inicio=True)
    print("")  # Nova linha após chamadas recursivas

    for value in graph.vertices.values():
        print("---", value)

    print("3 - Resetando o estado dos nos")
    graph.restaurar_nos()

    print("4 - Explorando o labirinto (RAM) com o BFS")
    indice_final = graph.bfs(indice_inicial)

    print("5 - Encontrando o menor caminho")
    menor_caminho = graph.encontrar_caminho(indice_final)
    print("---", menor_caminho)

    print("6 - Validando o menor caminho encontrado")
    api.validar_caminho(menor_caminho)

    timer_main.parar()

    print("7 - Estatisticas finais")
    print("--- API Calls:", api.n_calls)
    print("--- Tempo total do programa   : %.3f segundos" % timer_main.tempo_total)
    print("--- Tempo total das API Calls : %.3f segundos" % api.timer.tempo_total)
    proporcao = (api.timer.tempo_total * 100) / timer_main.tempo_total
    print("--- (%.2f%% do total do programa)" % proporcao)

File: src/solver.rs
use core::panic;

use crate::api::API;
use crate::graph::Graph;
use crate::timer::Timer;
use crate::strategy::Strategy;

pub fn solver(address: Option<String>, maze: Option<String>, strategy: Option<Strategy>, custom_end: Option<i32>) {
    let mut timer = Timer::new();
    let mut api = API::novo(address, maze);
    let mut graph = Graph::novo();
    
    println!("I - Initial API call (URL: {} | Maze: {} | Strategy: {:?})", api.url, api.maze, strategy);
    let first_pos = api.start(&mut graph.vertices, custom_end).unwrap();

    let path: Vec<i32> = match strategy {
        None => {
            panic!("No strategy selected!");
        }

        Some(strat) => match strat {
            Strategy::DFSFullBFS => {
                println!("2.1 Explorando todo o labirinto com o DFS");
                //graph.dfs_total(&mut api, indice_inicial, -1, true); 
                graph.dfs_otimizado(&mut api, first_pos, false);
                println!(""); // Nova linha após chamadas recursivas
    
                println!("2.2 Resetando o estado dos nos");
                graph.restaurar_nos();
    
                println!("2.3 Encontrando o menor caminho com o BFS");
                //graph.bfs_ram(indice_inicial)
                graph.bfs_restricted(first_pos, graph.indice_final.unwrap())
            }
    
            Strategy::DFSExitBFS => {
                println!("2.1 Explorando o labirinto ate o alvo com o DFS");
                //graph.dfs_alvo(&mut api, indice_inicial, -1, true);
                graph.dfs_otimizado(&mut api, first_pos, true);
                println!(""); // Nova linha após chamadas recursivas
                
                println!("2.2 Resetando o estado dos nos");
                graph.restaurar_nos();
    
                println!("2.3 Encontrando o menor caminho com o BFS");
                //graph.bfs_ram(indice_inicial)
                graph.bfs_restricted(first_pos, graph.indice_final.unwrap())
            }
    
            Strategy::DFS => {
                println!("2 - Explorando o labirinto ate o alvo com o DFS");
                //graph.dfs_alvo(&mut api, indice_inicial, -1, true);
                graph.dfs_otimizado(&mut api, first_pos, true);
                println!(""); // Nova linha após chamadas recursivas
                graph.encontrar_caminho(graph.indice_final.unwrap())
            }
    
            Strategy::BFS => {
                println!("2 - Explorando o labirinto ate o alvo com o BFS");
                let aux = graph.bfs_explorer(&mut api, first_pos);
                println!("");
                graph.encontrar_caminho(aux)
            }
        }
    };

    println!("3 - Gerando a lista com o caminho encontrado");
    println!("--- {:?}", path);

    println!("4 - Validating found path");
    api.validate_path(path);

    timer.stop();

    println!("5 - Final statistics");
    println!("--- API Calls: {}", api.n_calls);
    println!("--- Explored nodes: {}", graph.vertices.values().len());
    println!("--- Total running time  : {:.3}", timer.total());
    println!("--- Time spent on API Calls: {:.3}", api.timer.total());
    let ratio: f64 = (api.timer.total() * 100.0) / timer.total();
    println!("--- ({:.2}% of running time)", ratio);
}